src/algebra/pointwise.lean

```diff
@@ -880,10 +880,35 @@ by { simp only [← image2_mul, image_image2, image2_image_left, image2_image_ri
 lemma preimage_mul_preimage_subset {s t : set β} : m ⁻¹' s * m ⁻¹' t ⊆ m ⁻¹' (s * t) :=
 by { rintros _ ⟨_, _, _, _, rfl⟩, exact ⟨_, _, ‹_›, ‹_›, (m.map_mul _ _).symm ⟩ }
 
+instance set_semiring.no_zero_divisors : no_zero_divisors (set_semiring α) :=
+{ eq_zero_or_eq_zero_of_mul_eq_zero := λ a b ab, by
+  { by_cases a0 : a = 0,
+  { exact or.inl a0 },
```

Comment:
```suggestion
  { refine a.eq_empty_or_nonempty.imp_right (λ ha, b.nonempty_or_eq_empty.symm.imp_right $ λ hb, _),
```
I leave you the pleasure of making it a one-liner from there 😉

Reply:
I had some difficulty making this work, but I managed to find a term-mode proof: is this what you had in mind?

Reply:
See below 😉